### PR TITLE
[rust] basic oneOf support

### PR DESCRIPTION
**Pull request by ctrlaltf24**
Generates oneOf enums.

```yaml
openapi: 3.0.1

info:
  version: "1.0.0"
  title: rust one-of fix
paths: {}
components:
  schemas:
    Foo:
      x-one-of-name: Foo
      oneOf:
        - type: object
          required:
            - field_one
          properties:
            field_one:
              type: string
            field_two:
              type: string
        - type: object
          required:
            - field_three
          properties:
            field_three:
              type: number
```

Expected behavior: Able to fill in struct fields in such a way that creates a valid request
Actual behavior (on master): Impossible to fill the struct Foo to make it a valid schema (as BOTH required fields (field_one and field_three) are required to be filled in rust, yet the openapi schema only allows one or the other to be set.
Actual behavior (on this PR): able to fill fields by using the corresponding variant on the enum Foo.

```rust
use super::FooOneOf;
use super::FooOneOf1;

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum Foo {
    FooOneOf(Box<FooOneOf>),
    FooOneOf1(Box<FooOneOf1>),
}
```

### Known Limitations
- model.mds still need to be updated to match the new enum scheme

### Related issue
https://github.com/OpenAPITools/openapi-generator/issues/9497

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@frol @farcaller @richardwhiuk @paladinzh @jacob-pro

### CODE REVIEW

**Comment by wing328** [comment]
@nshaaban-cPacket thanks for the PR. Can you please PM me via Slack when you've time? 

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

**Comment by nshaaban-cPacket** [comment]
kk talked to @wing328 :
- Pushing support oneOf and properties support out to a later date/PR
- Pushing discriminator support out to a later date
- Will implement some tests https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests to make sure this works, and stays working.
- Double check this PR will work with my spec. @mattoni feel free to test as well. May push further fixes to other PRs as needed

**Comment by wing328** [comment]
@nshaaban-cPacket is this PR ready for another review? 

cc @frol @farcaller @richardwhiuk @paladinzh @jacob-pro

**Comment by nshaaban-cPacket** [comment]
> @nshaaban-cPacket is this PR ready for another review?

yup!

**Comment by wing328** [comment]
right, the circleci node1 failure can be ignore.

**Comment by mattoni** [comment]
Is there anything still blocking this being merged? Excited to have this in :)

**Comment by nshaaban-cPacket** [comment]
> Is there anything still blocking this being merged? Excited to have this in :)

Just waiting on review from some of the rust technical community. Feel free to give a try yourself in the meantime!

**Comment by wing328** [comment]
@nshaaban-cPacket thanks for the PR. Is it correct to say that the oneOf implementation uses Rust's enum?

**Comment by nshaaban-cPacket** [comment]
> @nshaaban-cPacket thanks for the PR. Is it correct to say that the oneOf implementation uses Rust's enum?

Correct, this PR uses an enum for oneOf. Current version doesn't use a struct (similar to a class in other languages) with optional fields

**Comment by wing328** [comment]
@nshaaban-cPacket can it handle the following?
```
    RgbColor:
      description: RGB three element array with values 0-255.
      type: array
      items:
        type: integer
        minimum: 0
        maximum: 255
      minItems: 3
      maxItems: 3
    RgbaColor:
      description: RGBA four element array with values 0-255.
      type: array
      items:
        type: integer
        minimum: 0
        maximum: 255
      minItems: 4
      maxItems: 4
    HexColor:
      description: 'Hex color string, such as #00FF00.'
      type: string
      pattern: ^#(?:[0-9a-fA-F]{3}){1,2}$
      minLength: 7
      maxLength: 7
    Color:
      description: RGB array, RGBA array, or hex string.
      oneOf:
      - $ref: '#/components/schemas/RgbColor'
      - $ref: '#/components/schemas/RgbaColor'
      - $ref: '#/components/schemas/HexColor'


```

File: temp.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+openapi: 3.0.1
```
Comment (frewsxcv) [inline temp.yaml:1]: Did you mean to commit this file?

**Comment by ctrlaltf24** [comment]
> @nshaaban-cPacket can it handle the following?

Given a config of
```
generatorName: rust
outputDir: samples/client/others/rust/hyper/oneOf-color
library: hyper
inputSpec: modules/openapi-generator/src/test/resources/3_0/rust/color.yaml
templateDir: modules/openapi-generator/src/main/resources/rust
generateAliasAsModel: true
additionalProperties:
  supportAsync: "false"
  packageName: oneof-color-hyper
```

And a openapi of
```
openapi: 3.0.0
servers:
  - url: 'http://localhost'
info:
  description: >-
    This is a simple sample for oneOf support
  version: 1.0.0
  title: OpenAPI Color
  license:
    name: Apache-2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0.html'
paths:
  /color:
    get:
      summary: Get the current color
      description: ''
      operationId: getColor
      responses:
        '200':
          description: successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Color'
components:
  schemas:
    RgbColor:
      description: RGB three element array with values 0-255.
      type: array
      items:
        type: integer
        minimum: 0
        maximum: 255
      minItems: 3
      maxItems: 3
    RgbaColor:
      description: RGBA four element array with values 0-255.
      type: array
      items:
        type: integer
        minimum: 0
        maximum: 255
      minItems: 4
      maxItems: 4
    HexColor:
      description: 'Hex color string, such as #00FF00.'
      type: string
      pattern: ^#(?:[0-9a-fA-F]{3}){1,2}$
      minLength: 7
      maxLength: 7
    Color:
      description: RGB array, RGBA array, or hex string.
      oneOf:
      - $ref: '#/components/schemas/RgbColor'
      - $ref: '#/components/schemas/RgbaColor'
      - $ref: '#/components/schemas/HexColor'
```

It correctly generates the Color model as the following
```rust
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Color {
    /// RGB array, RGBA array, or hex string.
    RgbColor(Box<RgbColor>),
    /// RGB array, RGBA array, or hex string.
    RgbaColor(Box<RgbaColor>),
    /// RGB array, RGBA array, or hex string.
    String(Box<String>),
}
```

However the current state of the rust generator (as it exists both on master and this PR) fails to generate the correct model for the aliases. This issue exists on master as well, and is out of scope of the PR to fix. A separate PR could be created to fix this, and in combination with this PR would result in complete support for the given spec.

```
#[derive(Clone, Debug, PartialEq, Default, Serialize, Deserialize)]
pub struct RgbColor {
}
```

**Comment by wing328** [comment]
@ctrlaltf24 can you please resolve the merge conflicts when you've time?

there's been change to the rust client generator so not sure if it has addressed the existing limitations you mentioned above.

**Comment by humb1t** [comment]
@ctrlaltf24 I made a PR with rebased master to your fork, hope it can help https://github.com/ctrlaltf24/openapi-generator/pull/1

**Comment by humb1t** [comment]
@wing328 - @ctrlaltf24 merged rebase onto master branch in his fork

**Comment by wing328** [comment]
I also did a test to manually trigger a build failure and the workflow caught the issue as expected: https://github.com/OpenAPITools/openapi-generator/actions/runs/7885693679/job/21517363148?pr=17854

let's give this PR a try.

thanks again for the contribution.

**Pull request merged**